File: 6/datum/datum_brut.cpp
#include <bits/stdc++.h>

using namespace std;

vector<int> dates;
int q;
string date;

int mj[12] = {31,29,31,30,31,30,31,31,30,31,30,31};

bool palin(int y)
{
	int day = (y % 10) * 10 + (y % 100) / 10;
	int month = y / 1000 + ((y / 100) % 10) * 10;
	return (day <= mj[month - 1] && day != 0 && month != 0 && month <= 12);
}

string ispis(int y)
{
	string ret = "";
	ret += (y % 10 + '0');
	ret += (y / 10) % 10 + '0';
	ret += ".";
	ret += (y / 100) % 10 + '0';
	ret += y / 1000 + '0';
	ret += ".";
	ret += (y / 1000) + '0';
	ret += (y / 100) % 10 + '0';
	ret += (y / 10) % 10 + '0';
	ret += y % 10 + '0';
	
	return ret;
}

bool kasnije(string a, string b)
{
	int mja = (a[3]- '0') * 10 + a[4] - '0';
	int mjb = (b[3]- '0') * 10 + b[4] - '0';
	int daa = (a[0]- '0') * 10 + a[1] - '0';
	int dab = (b[0]- '0') * 10 + b[1] - '0';
	if(mja > mjb)
		return true;
	if(mja == mjb)
	{
		if(daa > dab)
			return true;
		return false;
	}
	return false;
}

int main()
{
	clock_t begin, end;
	double time_spent;
	begin = clock();
	for(int i = 1; i < 10000; i++)
		if(palin(i))
			dates.push_back(i);
	cin >> q;
	while(q--)
	{
		cin >> date;
		int year = (date[6] - '0') * 1000 + (date[7] - '0') * 100 + (date[8] - '0') * 10 + date[9] - '0';
		for(int i = 0; i < 366; i++)
		{
			if(year == dates[i])
			{
				string pal = ispis(year);
				if(kasnije(pal, date))
					cout << pal << "." << endl;
				else
					cout << ispis(dates[i + 1]) << "." << endl;
				break;
			}
			else if(year < dates[i])
			{
				cout << ispis(dates[i]) << "." << endl;
				break;
			}
		}
	}
	end = clock();
	time_spent = (double)(end - begin) / CLOCKS_PER_SEC;
	//cout << endl<< time_spent;
	return 0;
}

File: 6/datum/datum_brut.py
import time

dates = [];
mj= [31,29,31,30,31,30,31,31,30,31,30,31];

def palin(y):
	day = (y % 10) * 10 + (y % 100) / 10;
	month = y / 1000 + ((y / 100) % 10) * 10;
	return (month != 0 and month <= 12 and day <= mj[month - 1] and day != 0);

def ispis(y):
	ret = "";
	ret += str(y % 10)
	ret += str((y / 10) % 10)
	ret += "."
	ret += str((y / 100) % 10)
	ret += str(y / 1000)
	ret += ".";
	ret += str((y / 1000))
	ret += str((y / 100) % 10)
	ret += str((y / 10) % 10)
	ret += str(y % 10)
	return ret;

def bins(y):
	lo = 0
	hi = 365
	mi = -1
	while(lo < hi):
		mi = (hi + lo) // 2
		if(dates[mi] < y):
			lo = mi + 1
		else:
			hi = mi
	return dates[lo]

def nadi(y):
	lo = 0
	hi = 365
	mi = -1
	while(lo < hi):
		#print(mi, dates[mi])
		mi = lo + (hi - lo) // 2
		if(dates[mi] == y):
			return 1
		elif(dates[mi] < y):
			lo = mi + 1
		else:
			hi = mi
	return 0

def kasnije(a, b):
	mja = (ord(a[3])- nul) * 10 + ord(a[4]) - nul;
	mjb = (ord(b[3])- nul) * 10 + ord(b[4]) - nul;
	daa = (ord(a[0])- nul) * 10 + ord(a[1]) - nul;
	dab = (ord(b[0])- nul) * 10 + ord(b[1]) - nul;
	if(mja > mjb):
		return 1
	if(mja == mjb):
		if(daa > dab):
			return 1
		return 0
	return 0


start_time = time.time()

for i in range(10000):
	if(palin(i)):
		dates.append(i);

q = int(input())
nul = ord('0')

while(q):
	q = q - 1
	date = raw_input()
	year = (ord(date[6]) - nul) * 1000 + (ord(date[7]) - nul) * 100 + (ord(date[8]) - nul) * 10 + ord(date[9]) - nul
	for i in range(366):
		if(year == dates[i]):
			pal = ispis(year)
			if(kasnije(pal, date)):
				print(pal)
			else:
				print(ispis(dates[i + 1]))
			break
		elif(year < dates[i]):
			print(ispis(dates[i]))
			break
print("--- %s seconds ---" % (time.time() - start_time))
